### PR TITLE
Replace instance variable access with call to .elf.

### PR DESCRIPTION
**Pull request by rmNULL**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

`.elf` was introduced in gem release v1.1.0.
Resolves  https://github.com/Homebrew/brew/pull/7678#discussion_r442090941

### CODE REVIEW

**Comment by rmNULL** [comment]
cc @sjackman @woodruffw

**Comment by jonchang** [comment]
We don’t have version requirements for the other gems. Should we update Gemfile.lock instead?

**Comment by woodruffw** [comment]
> We don’t have version requirements for the other gems. Should we update Gemfile.lock instead?

Perhaps both? We won't be able to function with versions of `patchelf.rb` below 1.1 from here on out, so it makes sense to me to enforce it in the `Gemfile` itself and *also* to check-in the relevant lockfile changes. 

@rmNULL, could you do that?

**Comment by jonchang** [comment]
> Perhaps both? We won't be able to function with versions of patchelf.rb below 1.1 from here on out, so it makes sense to me to enforce it in the Gemfile itself and also to check-in the relevant lockfile changes.

Yeah, but it should always update to the latest in the lockfile, since setting `HOMEBREW_PATCHELF_RB` will always call `install_bundler_gems!`, which will always install the version specified in the lockfile.

**Comment by sjackman** [comment]
For consistency with the rest of the `Gemfile` I agree with removing the version dependency.

**Comment by woodruffw** [comment]
> Yeah, but it should always update to the latest in the lockfile, since setting `HOMEBREW_PATCHELF_RB` will always call `install_bundler_gems!`, which will always install the version specified in the lockfile.
> 
> I note that patchelf isn't currently specified in `Gemfile.lock`, so this means that it'll always be up to date assuming it doesn't conflict with other things in the lockfile.

Good points. No objections here!

**Comment by jonchang** [comment]
If anything, we should:

* remove the ENV variable test in `Gemfile`
* commit `patchelf` to Gemfile.lock

This is fine because (and correct me if I'm wrong) we always check for `HOMEBREW_PATCHELF_RB` and call `install_vendor_gems` on all codepaths that would use `patchelf.rb`. This means that we no longer get dirty work trees when `HOMEBREW_PATCHELF_RB` is set, and anyone actually using the functionality will have the right gems installed on demand.

When we flip the switch so everyone uses `patchelf.rb`, we can use `brew vendor-gems` and commit `patchelf.rb` under `vendor/`.

**Comment by woodruffw** [comment]
> This is fine because (and correct me if I'm wrong) we always check for `HOMEBREW_PATCHELF_RB` and call `install_vendor_gems!` on all codepaths that would use `patchelf.rb`. This means that we no longer get dirty work trees when `HOMEBREW_PATCHELF_RB` is set, and anyone actually using the functionality will have the right gems installed on demand.
> 
> When we flip the switch so everyone uses `patchelf.rb`, we can use `brew vendor-gems` and commit `patchelf.rb` under `vendor/`.

Yeah, this sounds preferable to me. I can't recall why we didn't do this in the first place, other than maybe not wanting to install `patchelf.rb` unconditionally on all Homebrew boxes. But we'll be doing that when we vendor eventually, so I don't think it's a huge deal.

**Comment by jonchang** [comment]
Now that `HOMEBREW_DEVELOPER` implies `HOMEBREW_PATCHELF_RB` there's no reason to avoid committing to the lockfile anymore. But prior to that change I think it made sense.

**Comment by Bo98** [comment]
> Now that `HOMEBREW_DEVELOPER` implies `HOMEBREW_PATCHELF_RB` there's no reason to avoid committing to the lockfile anymore. But prior to that change I think it made sense.

Would it affect dependabot at all?

**Comment by jonchang** [comment]
> Would it affect Dependabot at all?

**Comment by Bo98** [comment]
I asked since Dependabot can purge dependencies no longer used, and I wasn't sure how it did that internally.

**Comment by jonchang** [comment]
I think dependabot basically does `bundle update foo` on the lockfile, so if we make the Gemfile more "standard" (not changing behavior based on environment variables) it'll make things easier for everyone, including the bots :)

**Comment by Bo98** [comment]
> make the Gemfile more "standard" (not changing behavior based on environment variables)

:+1: on this

**Comment by sjackman** [comment]
> Yeah, this sounds preferable to me. I can't recall why we didn't do this in the first place, other than maybe not wanting to install patchelf.rb unconditionally on all Homebrew boxes.

The conditional `Gemfile` was at Mike's request, and yes, that was the reasoning.

File: Library/Homebrew/Gemfile [hunk 1]
```diff
@@ -23,7 +23,7 @@ end
 gem "activesupport"
 gem "concurrent-ruby"
 gem "mechanize"
-gem "patchelf" if ENV["HOMEBREW_PATCHELF_RB"]
+gem "patchelf", "~> 1.1" if ENV["HOMEBREW_PATCHELF_RB"]
```
Comment (sjackman) [inline Library/Homebrew/Gemfile:26]:
```suggestion
gem "patchelf" if ENV["HOMEBREW_PATCHELF_RB"]
```

Comment (sjackman) [reply Library/Homebrew/Gemfile:26]: LGTM after this small change.

Comment (woodruffw) [reply Library/Homebrew/Gemfile:26]: I'm gonna go ahead and commit this, since it's already formatted as a suggestion.

**Pull request merged**